geomstats/stratified_geometry/wald_space.py

```diff
@@ -0,0 +1,1024 @@
+r"""Classes for the Wald Space and elements therein of class Wald and helper classes.
+
+Class ``Split``.
+Essentially, a ``Split`` is a two-set partition of a subset of :math:`\{0,\dots,n-1\}`.
+This class is designed such that one part of both parts of the partition can be empty.
+Splits are corresponding uniquely to edges in a phylogenetic forest, where, if one cuts
+the edge in the forest, the resulting two-set partition of the labels of the respective
+component of the forest is the corresponding split.
+
+Class ``Structure``.
+A structure is a partition into non-empty sets of the set :math:`\{0,\dots,n-1\}`,
+together with a set of splits for each element of the partition, where every split is a
+two-set partition of the respective element.
+A structure basically describes a phylogenetic forest, where each set of splits gives
+the structure of the tree with the labels of the corresponding element of the partition.
+
+Class ``Wald``.
+A wald is essentially a phylogenetic forest with weights between zero and one on the
+edges. The forest structure is stored as a ``Structure`` and the edge weights are an
+array of length that is equal to the total number of splits in the structure. These
+elements are the points in Wald space and other phylogenetic forest spaces, like BHV
+space, although the partition is just the whole set of labels in this case.
+
+Class ``WaldSpace``.
+A topological space. Points in Wald space are instances of the class :class:`Wald`:
+phylogenetic forests with edge weights between 0 and 1.
+In particular, Wald space is a stratified space, each stratum is called grove.
+The highest dimensional groves correspond to fully resolved or binary trees.
+The topology is obtained from embedding wälder into the ambient space of strictly
+positive :math:`n\times n` symmetric matrices, implemented in the
+class :class:`spd.SPDMatrices`.
+
+
+Lead author: Jonas Lueg
+
+
+References
+----------
+[Garba21]_  Garba, M. K., T. M. W. Nye, J. Lueg and S. F. Huckemann.
+            "Information geometry for phylogenetic trees"
+            Journal of Mathematical Biology, 82(3):19, February 2021a.
+            https://doi.org/10.1007/s00285-021-01553-x.
+[Lueg21]_   Lueg, J., M. K. Garba, T. M. W. Nye, S. F. Huckemann.
+            "Wald Space for Phylogenetic Trees."
+            Geometric Science of Information, Lecture Notes in Computer Science,
+            pages 710–717, Cham, 2021.
+            https://doi.org/10.1007/978-3-030-80209-7_76.
+"""
+
+
+import functools
+import itertools as it
+
+import numpy as np
```

Comment:
`import geomstats.backend as gs`

I know it only works with `np`, but maybe soon we will be able to extend to different backends (e.g. `jax`).

Reply:
Does this mean I should never import numpy?

Reply:
Exactly. `gs` will have all the functions and methods you need though.

Reply:
So, for example, it has ``gs.maximum``, but not ``gs.minimum``.
Should I

1. ask you to add it to the backends;
2. add it myself to the backends;
3. add it myself only to the numpy backend;
4. or try to find a workaround (for example ``-gs.maximum(-x) == gs.minimum``?

Reply:
`gs.minimum` is now available (see [https://github.com/geomstats/geomstats/blob/master/geomstats/_backend/numpy/__init__.py]). I'll see what I can do for `np.fill_diagonal`.

Reply:
thank you :)